protocol/msgpack/types.go

```diff
@@ -326,11 +326,22 @@ type AggregatedEncoder interface {
 	// EncodeMetricWithStoragePolicy encodes a metric with an applicable storage policy.
 	EncodeMetricWithStoragePolicy(mp aggregated.MetricWithStoragePolicy) error
 
+	// EncodeMetricWithStoragePolicyAndEncodeTime encodes a metric with an applicable
+	// storage policy, alongside the time at which encoding happens.
+	EncodeMetricWithStoragePolicyAndEncodeTime(
```

Comment:
Do we see ourselves adding other fields in the future? Seems a little heavyweight to have to add a new function each time. Perhaps it would be worth refactoring so that we only have a single function: `EncodeMetric` which takes a `MetricUnion`, a struct consisting of an aggregated metric, a chunked metric, a storage policy, an encoded time, and a byte which indicates the type of the union: e.g. `MetricWithStoragePolicy`, `MetricWithStoragePolicyAndEncodeTime`, etc.

Reply:
I generally prefer to keep the API as simple as possible and avoid union types in the API. This way if the user pass in the wrong type, they'll get compilation errors, as opposed to passing in the union with the wrong type set in the union which will be a logic error and much harder to debug.

I actually don't think it's too big of an issue to introduce a new API when the encoder supports encoding a new type (metricWithStoragePolicy combined with the encoding time is essentially a new type without being explicitly defined). After all, the encoded data will be type-specific.

Reply:
Fair point, I'll also take compilation guarantees over API simplicity. I've never felt strongly about Go supporting abstract data types, but language support for union types here would be incredibly helpful.

Reply:
Yeah using union types is basically trading some amount of type safety for slightly less verbose implementation. Generics would give us the better of both worlds but..